### PR TITLE
Return the environment variables currently known to shell

### PR DESCRIPTION
**Pull request by bzm3r**
Addresses #77 

TO DETERMINE:

- [ ] is it really necessary to have both `vars` and `vars_os`?
- [x] what should the return type of `vars`/`vars_os` be? For now, I defaulted to a "minimal iterable which does not retain references to the actual data" (a `Vec`) ([resolved](https://github.com/matklad/xshell/pull/78#issuecomment-1881071175))

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -662,6 +662,33 @@ impl Shell {
...
+    /// The returned hash map contains a snapshot of the process’s environment
+    /// variables at the time of this invocation. Modifications to environment
+    /// variables afterwards will not be reflected in the returned iterator.
+    ///
+    /// # Panics
+    ///
+    /// While iterating, the returned iterator will panic if any key or value in
+    /// the environment is not valid unicode. If this is not desired, consider
+    /// using [`Shell::vars_os`](Self::vars_os).
+    pub fn vars(&self) -> Vec<(String, String)> {
```
Comment (lnicola) [inline src/lib.rs:684]: We should filter them out instead of panicking, I think that's also what the std one does.

Comment (bzm3r) [reply src/lib.rs:684]: @lnicola Surprisingly: https://doc.rust-lang.org/stable/std/env/fn.vars.html#panics

Comment (bzm3r) [reply src/lib.rs:684]: I guess filtering has the difficulty of hiding something the user might want to be aware of? Perhaps we ought to return a `Result` then?

Comment (lnicola) [reply src/lib.rs:684]: Oh, sorry, for some reason I thought it did that. It even made sense in a way 😀.

**Comment by matklad** [comment]
Yes, this makes sense to me! I paused a bit on the allocating API, but I think that's fine --- the std forces an allocation as well, and, while it has a separate itearator, I don't think that's going to be much of a win over returning a vector. 

However, let's sort the vector before returning, to increase determinism, like we do in `read_dir`. We _don't_ aim to be zero-overhead, but we do want to be robust, and being deterministic is pretty important for robustness.

**Comment by bzm3r** [comment]
@matklad I usually use `sort_unstable_by`, but given that our concern is determinism, perhaps that is the wrong move here? However, there are no duplicate keys (it is a HashMap that we are returning as an iterable, after all), so the warnings associated with `sort_unstable_by` should not apply when sorting by the key element of the tuples? Can you confirm?

**Comment by matklad** [comment]
unstable sort is fine --- the keys are guaranteed to be unique. We might actually go and pass a custom comparison function, which only compares first elements of the tuples, to reduce the code bloat in a minor way, as we know that comparing the second element will never be necessary.

**Comment by matklad** [comment]
This now exists on master, thanks for the pull request!

**Pull request closed without merging**